File: kqt20/41_First_Missing_positive.java
class Solution {
    public int firstMissingPositive(int[] nums) {

        int min = 1;
        int max = nums.length;

        int tmp;

        int result;

        for(int i=0; i<max; i++){
            while(nums[i] >= min && nums[i] <= max && nums[nums[i]-1] != nums[i]){
                tmp = nums[nums[i]-1];
                nums[nums[i]-1] = nums[i];
                nums[i] = tmp;
            }
        }
        for(int i=0; i<max; i++){
            if(nums[i] < min || nums[i] > max || nums[i] != i+1){
                result = i+1;
                return result;
            }
        }
        result = max+1;
        return result;
    }
}

File: onestep00/41_First_Missing_positive.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        nums.append(0)
        l = len(nums)
        
        for i in range(l):
            if nums[i] < 0 or nums[i] >= l:
                nums[i] = 0
                
        for i in range(l):
            nums[nums[i] % l] += l
        
        for i in range(l):
            if nums[i]/l < 1:
                return i
            
        return l
